### PR TITLE
fix(virtual-scroll): emit on `viewChange` inside the `NgZone`

### LINKED ISSUE #12869: virtual-scroll: Change detection not triggered when using DataSource that uses viewChange.
#### Bug, feature request, or proposal:
Change detection is not triggered when you scroll if you are using a custom DataSource that utilizes viewChange.  I have no idea what the appropriate way to fix this is, but this is what I did and it solved my problem:

```
const onRenderedDataChange = CdkVirtualForOf.prototype['_onRenderedDataChange'];
CdkVirtualForOf.prototype['_onRenderedDataChange'] = function (this: any) {
    onRenderedDataChange.apply(this, arguments);
    if (!this._changeDetectorRef) {
        this._changeDetectorRef = this._viewContainerRef.injector.get(ChangeDetectorRef);
    }
    this._changeDetectorRef.detectChanges();
}
```
#### What is the expected behavior?
Change detection should occur when the underlying DataSource's data changes.

#### What is the current behavior?
Change detection does not occur when the underlying DataSource's data changes.

#### What are the steps to reproduce?
This demo is almost identical to the one that I created for #11289.  In that demo I used setInterval(noop, 1000) to suppress this problem.  That is obviously a horrible solution, but I was trying to focus on that particular bug.  In this demo, just scroll to the bottom and the words "Loading..." will not disappear like they should.  If you put setInterval(noop, 1000) in the constructor or apply my patch mentioned above, the problem will go away.

https://stackblitz.com/edit/angular-material2-issue-czyxyd

#### What is the use-case or motivation for changing an existing behavior?
It's broken.

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
@angular/cdk-experimental 6.4.6

#### Is there anything else we should know?
No.

### PR DESCRIPTION
**Pull request by mmalerba**
fixes #12869

### CODE REVIEW

File: src/cdk/scrolling/virtual-for-of.ts [hunk 1]
```diff
@@ -166,14 +167,16 @@ export class CdkVirtualForOf<T> implements CollectionViewer, DoCheck, OnDestroy
       /** The set of available differs. */
       private _differs: IterableDiffers,
       /** The virtual scrolling viewport that these items are being rendered in. */
-      @SkipSelf() private _viewport: CdkVirtualScrollViewport) {
+      @SkipSelf() private _viewport: CdkVirtualScrollViewport,
+      /** The Angular zone. */
+      ngZone: NgZone) {
```
Comment (crisbeto) [inline src/cdk/scrolling/virtual-for-of.ts]: Should this be optional with a breaking change for next major?

File: src/cdk/scrolling/virtual-for-of.ts [hunk 2]
```diff
@@ -166,14 +167,16 @@ export class CdkVirtualForOf<T> implements CollectionViewer, DoCheck, OnDestroy
       /** The set of available differs. */
       private _differs: IterableDiffers,
       /** The virtual scrolling viewport that these items are being rendered in. */
-      @SkipSelf() private _viewport: CdkVirtualScrollViewport) {
+      @SkipSelf() private _viewport: CdkVirtualScrollViewport,
+      /** The Angular zone. */
```
Comment (crisbeto) [inline src/cdk/scrolling/virtual-for-of.ts]: This comment doesn't contribute much.

File: src/cdk/scrolling/virtual-for-of.ts (hunk 1 above)
Comment (mmalerba) [reply src/cdk/scrolling/virtual-for-of.ts]: virtual scroll isn't released yet so its ok

File: src/cdk/scrolling/virtual-for-of.ts (hunk 2 above)
Comment (mmalerba) [reply src/cdk/scrolling/virtual-for-of.ts]: ok, removed

File: src/cdk/scrolling/virtual-for-of.ts [hunk 3]
```diff
@@ -166,14 +167,15 @@ export class CdkVirtualForOf<T> implements CollectionViewer, DoCheck, OnDestroy
...
+      @SkipSelf() private _viewport: CdkVirtualScrollViewport,
+      ngZone: NgZone) {
     this.dataStream.subscribe(data => {
       this._data = data;
       this._onRenderedDataChange();
     });
     this._viewport.renderedRangeStream.pipe(takeUntil(this._destroyed)).subscribe(range => {
       this._renderedRange = range;
-      this.viewChange.next(this._renderedRange);
+      ngZone.run(() => this.viewChange.next(this._renderedRange));
```
Comment (crisbeto) [inline src/cdk/scrolling/virtual-for-of.ts]: You should be able to write a test for this. Something like:

```
const spy = jasmine.createSpy('something');

renderedRangeStream.subscribe(() => spy(NgZone.isInAngularZone()));

expect(spy).toHaveBeenCalledWith(true);
```

File: src/cdk/scrolling/virtual-for-of.ts [hunk 4]
```diff
@@ -166,14 +167,15 @@ export class CdkVirtualForOf<T> implements CollectionViewer, DoCheck, OnDestroy
       /** The set of available differs. */
       private _differs: IterableDiffers,
       /** The virtual scrolling viewport that these items are being rendered in. */
-      @SkipSelf() private _viewport: CdkVirtualScrollViewport) {
+      @SkipSelf() private _viewport: CdkVirtualScrollViewport,
+      ngZone: NgZone) {
```
Comment (crisbeto) [inline src/cdk/scrolling/virtual-for-of.ts]: I left this before, but I think it got lost between commits: shouldn't we make this an optional parameter with a breaking change for the next major?

Comment (mmalerba) [reply src/cdk/scrolling/virtual-for-of.ts]: virtual scroll isn't released yet, so it's not necessary

**Pull request merged**